### PR TITLE
feat(rpc): Add client wrapper

### LINKED ISSUE #1185: Add client wrapper

### LINKED ISSUE #1186: Add fraud service to rpc

### PR DESCRIPTION
**Pull request by distractedm1nd**
Based on #1175 
Closes #1185 
Closes #1186 

Also adds the basis for the new rpc tests:
1. Making sure all methods we expose over the API actually exist
2. Making sure all the types we return can be marshalled in and out of json
3. RPC specific special behaviors, like authentication

### CODE REVIEW

File: nodebuilder/fraud/service.go [hunk 1]
```diff
@@ -1,9 +1,21 @@
...
 
-// Module encompasses the behavior necessary to subscribe and broadcast
-// fraud proofs within the network.
+	"github.com/celestiaorg/celestia-node/fraud"
+)
+
+// API is a wrapper around Module for the RPC.
+// TODO(@distractedm1nd): These structs need to be autogenerated.
+type API struct {
+	Broadcast func(context.Context, fraud.Proof) error
```
Comment (vgonkivs) [inline nodebuilder/fraud/service.go:12]: Why do you need a broadcaster here?

File: nodebuilder/rpc_test.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	require.NoError(t, err)
+	defer closer()
+	ctx := context.Background()
+
+	expectedBalance := &state.Balance{
+		Amount: sdk.NewInt(100),
+		Denom:  "utia",
+	}
+
+	server.EXPECT().Balance(gomock.Any()).Return(expectedBalance, nil).Times(1)
```
Comment (renaynay) [inline nodebuilder/rpc_test.go:35]: amazing!

File: api/rpc/client/client.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+	"github.com/celestiaorg/celestia-node/nodebuilder/daser"
+	"github.com/celestiaorg/celestia-node/nodebuilder/fraud"
+	"github.com/celestiaorg/celestia-node/nodebuilder/header"
+	"github.com/celestiaorg/celestia-node/nodebuilder/share"
+	"github.com/celestiaorg/celestia-node/nodebuilder/state"
+)
+
+//go:generate go run github.com/golang/mock/mockgen -destination=../../mocks/api.go -package=mocks . API
+type API interface {
```
Comment (renaynay) [inline api/rpc/client/client.go:16]: this interface is generated for mocking? Can you store these in a separate file, maybe `mock.go` per package? I think that's more standard.

Comment (distractedm1nd) [reply api/rpc/client/client.go:16]: So, they are stored in a separate file - the `go:generate` uses this API interface to autogenerate the mocks used in `api/mocks/api.go`

File: nodebuilder/fraud/service.go (hunk 1 above)
Comment (distractedm1nd) [reply nodebuilder/fraud/service.go]: removed

File: api/rpc/client/client.go (hunk 3 above)
Comment (renaynay) [reply api/rpc/client/client.go:16]: ack, resolve.

File: api/rpc/client/client.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+// CloseAll closes all saved clients.
+func (m *MultiClientCloser) CloseAll() {
+	for _, closer := range m.closers {
+		closer()
+	}
+}
+
+// NewClient creates a new Client with one connection per namespace.
+func NewClient(ctx context.Context, addr string) (*Client, *MultiClientCloser, error) {
```
Comment (Wondertan) [inline api/rpc/client/client.go:58]: As discussed, we can remove the Close and close all the namespaces at once in the Client.Close method

File: nodebuilder/das/service.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+)
+
+// API is a wrapper around Module for the RPC.
+// TODO(@distractedm1nd): These structs need to be autogenerated.
+type API struct {
+	SamplingStats func(ctx context.Context) (das.SamplingStats, error)
+}
+
+type Module interface {
+	SamplingStats(ctx context.Context) (das.SamplingStats, error)
```
Comment (renaynay) [inline nodebuilder/das/service.go:16]: this needs doc

File: nodebuilder/header/service.go [hunk 6]
```diff
@@ -8,6 +8,20 @@ import (
...
+// API is a wrapper around Module for the RPC.
+// TODO(@distractedm1nd): These structs need to be autogenerated.
+type API struct {
+	// GetByHeight returns the ExtendedHeader at the given height, blocking
+	// until header has been processed by the store or context deadline is exceeded.
+	GetByHeight func(context.Context, uint64) (*header.ExtendedHeader, error)
+	// Head returns the ExtendedHeader of the chain head.
+	Head func(context.Context) (*header.ExtendedHeader, error)
+	// IsSyncing returns the status of sync
+	IsSyncing func() bool
```
Comment (renaynay) [inline nodebuilder/header/service.go:20]: we don't need double docs between API struct and Module

I prefer docs on actual Module

File: nodebuilder/das/service.go [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package das
+
+import (
+	"context"
+
+	"github.com/celestiaorg/celestia-node/das"
+)
+
+// API is a wrapper around Module for the RPC.
+// TODO(@distractedm1nd): These structs need to be autogenerated.
```
Comment (Wondertan) [inline nodebuilder/das/service.go:10]: AN issue

File: nodebuilder/rpc/rpc.go [hunk 8]
```diff
@@ -12,14 +13,18 @@ import (
...
-	daser *das.DASer,
 ) {
 	serv.RegisterService("state", state)
 	serv.RegisterService("share", share)
+	serv.RegisterService("fraud", fraud)
 	serv.RegisterService("header", header)
-	serv.RegisterService("daser", daser)
+	if daser != nil {
+		serv.RegisterService("das", daser)
+	}
```
Comment (renaynay) [inline nodebuilder/rpc/rpc.go:27]: won't need this with #1299

File: nodebuilder/header/service.go [hunk 9]
```diff
@@ -8,6 +8,20 @@ import (
...
+}
```
Comment (Wondertan) [inline nodebuilder/header/service.go:21]: No need for docs, also move it before Module, or even down in the flow

File: nodebuilder/testing.go [hunk 10]
```diff
@@ -46,7 +46,7 @@ func TestNodeWithConfig(t *testing.T, tp node.Type, cfg *Config, opts ...fx.Opti
 	require.NoError(t, err)
 	cfg.Core.IP = ip
 	cfg.Core.RPCPort = port
-	cfg.RPC.Port = "0"
+	cfg.RPC.Port = "26655"
```
Comment (renaynay) [inline nodebuilder/testing.go:49]: y

Comment (renaynay) [reply nodebuilder/testing.go:49]: i forgot

**Review by renaynay — APPROVED** [review]

File: nodebuilder/testing.go (hunk 10 above)
Comment (Wondertan) [inline nodebuilder/testing.go:49]: ?

Comment (Wondertan) [reply nodebuilder/testing.go:49]: 25565

Comment (Wondertan) [reply nodebuilder/testing.go:49]: https://www.quora.com/Why-did-Minecraft-choose-port-25565-as-default

File: nodebuilder/header/service.go (hunk 9 above)
Comment (Wondertan) [reply nodebuilder/header/service.go:21]: Same for other modules

**Review by renaynay — COMMENTED** [review]
t.Cleanup closer and contexts in tests

File: nodebuilder/rpc_test.go [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+			impl := module.Type.Field(j)
+			method, _ := api.MethodByName(impl.Name)
+			require.Equal(t, method.Type, impl.Type)
+		}
+	}
+}
+
+// TODO(@distractedm1nd): Blocked by issues #1208 and #1207
+func TestAllReturnValuesAreMarshalable(t *testing.T) {
+	t.Skip()
```
Comment (Wondertan) [inline nodebuilder/rpc_test.go:56]: An issue to unskip this

File: nodebuilder/rpc_test.go [hunk 12]
```diff
@@ -0,0 +1,121 @@
+package nodebuilder
```
Comment (Wondertan) [inline nodebuilder/rpc_test.go:1]: Move to api

File: nodebuilder/rpc_test.go [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+	require.NoError(t, err)
+	defer closer.CloseAll()
+	ctx := context.Background()
+
+	expectedBalance := &state.Balance{
+		Amount: sdk.NewInt(100),
+		Denom:  "utia",
+	}
+
+	server.EXPECT().Balance(gomock.Any()).Return(expectedBalance, nil).Times(1)
```
Comment (Wondertan) [inline nodebuilder/rpc_test.go:34]: It's not worth having huge API mock just for a one line if we don't plan to add more similar tests

File: nodebuilder/testing.go (hunk 10 above)
Comment (distractedm1nd) [reply nodebuilder/testing.go:49]: will be resolved with #1286

Comment (distractedm1nd) [reply nodebuilder/testing.go:49]: duplicate of https://github.com/celestiaorg/celestia-node/pull/1195#discussion_r1009154625, resolving

File: nodebuilder/rpc_test.go (hunk 12 above)
Comment (distractedm1nd) [reply nodebuilder/rpc_test.go]: done

File: api/rpc/client/client.go (hunk 4 above)
Comment (distractedm1nd) [reply api/rpc/client/client.go]: done

File: nodebuilder/header/service.go (hunk 6 above)
Comment (distractedm1nd) [reply nodebuilder/header/service.go]: done

File: nodebuilder/das/service.go (hunk 5 above)
Comment (distractedm1nd) [reply nodebuilder/das/service.go]: done

File: nodebuilder/das/service.go (hunk 7 above)
Comment (distractedm1nd) [reply nodebuilder/das/service.go:15]: #1302

File: nodebuilder/rpc_test.go (hunk 11 above)
Comment (distractedm1nd) [reply nodebuilder/rpc_test.go]: #1303

**Review by renaynay — APPROVED** [review]
Approving, unit test failure is known flake that will be resolved with #1286

**Review by Wondertan — APPROVED** [review]

**Pull request merged**